src/main/java/com/minecolonies/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -340,21 +393,99 @@ protected int getMaxAttacksUntilRestock()
     protected AIState patrol()
     {
         worker.setAIMoveSpeed(1);
+        final AbstractBuilding building = getOwnBuilding();
 
-        if (currentPatrolTarget == null)
+        if (building instanceof BuildingGuardTower)
         {
-            currentPatrolTarget = getRandomBuilding();
+            if (currentPathTarget == null
+                    || BlockPosUtil.getDistance2D(building.getColony().getCenter(), currentPathTarget) > Configurations.workingRangeTownHall + Configurations.townHallPadding
+                    || currentPathTarget.getY() < 2)
+            {
+                return getNextPatrollingTarget((BuildingGuardTower) building);
+            }
+
+            if (worker.isWorkerAtSiteWithMove(currentPathTarget, PATH_CLOSE) || ((BuildingGuardTower) building).getTask().equals(BuildingGuardTower.Task.FOLLOW))
+            {
+                return getNextPatrollingTarget((BuildingGuardTower) building);
+            }
         }
 
-        if (worker.isWorkerAtSiteWithMove(currentPatrolTarget, PATH_CLOSE))
+        return AIState.GUARD_SEARCH_TARGET;
+    }
+
+    /**
+     * Retrieves the next patrolling target from the guard tower.
+     * @param building his building.
+     * @return the next state to go to.
+     */
+    private AIState getNextPatrollingTarget(final BuildingGuardTower building)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "getNextPatrollingTarget" is 14 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)